src/lib.rs

```diff
@@ -1,98 +1,53 @@
-use sha2::{Digest, Sha256};
-use std::sync::LazyLock;
-
-/// Length of a SHA256 hash in bytes.
-pub const HASH_LEN: usize = 32;
-
-/// Returns the digest of `input`.
-pub fn hash(input: &[u8]) -> Vec<u8> {
-    let mut hasher = Sha256::new();
-    hasher.update(input);
-    hasher.finalize().to_vec()
-}
+#![allow(dead_code)]
+#[cfg(feature = "poseidon2")]
+mod poseidon2;
+mod sha2;
 
-/// Hash function returning a fixed-size array (to save on allocations).
-pub fn hash_fixed(input: &[u8]) -> [u8; HASH_LEN] {
-    let mut hasher = Sha256::new();
-    hasher.update(input);
-    hasher.finalize().into()
-}
+/// Export
 
-/// Compute the hash of two slices concatenated.
-pub fn hash32_concat(h1: &[u8], h2: &[u8]) -> [u8; 32] {
-    let mut hasher = Sha256::new();
-    hasher.update(h1);
-    hasher.update(h2);
-    hasher.finalize().into()
-}
+#[cfg(feature = "sha2")]
+pub use crate::sha2::{hash, hash32_concat, hash_fixed, Context};
 
-pub trait Sha256Context {
-    fn new() -> Self;
+#[cfg(feature = "poseidon2")]
+pub use crate::poseidon2::{hash, hash32_concat, hash_fixed, Context};
 
-    fn update(&mut self, bytes: &[u8]);
-
-    fn finalize(self) -> [u8; HASH_LEN];
-}
-
-/// Context for incremental hashing
-pub struct Context(Sha256);
-
-impl Context {
-    pub fn new() -> Self {
-        Self(Sha256::new())
-    }
-
-    pub fn update(&mut self, bytes: &[u8]) {
-        self.0.update(bytes);
-    }
-
-    pub fn finalize(self) -> [u8; HASH_LEN] {
-        self.0.finalize().into()
-    }
-}
+/// Length of a SHA256 hash in bytes.
+pub const HASH_LEN: usize = 32;
 
 /// The max index that can be used with `ZERO_HASHES`.
 #[cfg(feature = "zero_hash_cache")]
 pub const ZERO_HASHES_MAX_INDEX: usize = 48;
 
+use std::sync::LazyLock;
 #[cfg(feature = "zero_hash_cache")]
 /// Cached zero hashes where `ZERO_HASHES[i]` is the hash of a Merkle tree with 2^i zero leaves.
 pub static ZERO_HASHES: LazyLock<Vec<[u8; HASH_LEN]>> = LazyLock::new(|| {
     let mut hashes = vec![[0; HASH_LEN]; ZERO_HASHES_MAX_INDEX + 1];
 
     for i in 0..ZERO_HASHES_MAX_INDEX {
-        hashes[i + 1] = hash32_concat(&hashes[i], &hashes[i]);
+        hashes[i + 1] = crate::hash32_concat(&hashes[i], &hashes[i]);
```

Comment:
```suggestion
        hashes[i + 1] = hash32_concat(&hashes[i], &hashes[i]);
```

`crate::` is unnecessary imo